torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -157,6 +157,12 @@ void ProcessGroupAgent::join() {
       SendWork(workerIds_[dst], Message({}, {}, MessageType::SHUTDOWN)));
   threadPool_.waitWorkComplete();
   listenerThread_.join();
+  // explicitly clean up python objects in PythonRpcHandler before destructing
+  // ProcessGroupAgent. Otherwise, in Python 3.5, when python program exits,
+  // the python objects in PythonRpcHandler could be cleaned up before
+  // destructing ProcessGroupAgent, then ProcessGroupAgent destructor will
+  // dec_ref deallocated python objects and crash.
+  PythonRpcHandler::getInstance().cleanUp();
```

Comment:
What if we just set `runUDFFunction_` and `loadResultFunction_` to None in the destructor of PythonRPCHandler? Does that work?

Reply:
if we only set them to None, not free the memory, then we are assuming python objects will be cleaned up before destructing pythonRpcHandler, but this assumption is only true for python 3.5. For python 3.6, it is not the case, we may be at risk of leaking memory in python 3.6 if set them as None only in destructor of PythonRPCHandler

Reply:
IIUC, PythonRpcHandler is a singleton so it would be destructed only on program exit. As a result, we won't leak any memory since the OS will just clean up all the memory once the process dies. Although, I'm not sure if ASAN would fail due to this. Can you give it a shot?

Reply:
@pritamdamania87 tests passed in both py3.5 and py3.6, fork asan did not complain as well, not sure whether spawn asan will complain or not because we disabled spawn asan. 

I will update diff to see how others think about it as well.